### ISSUE TITLE
item.gd and changes to gdscript in 3.1

### ISSUE THREAD

**Issue by santouits** [godotengine/escoria#237]
So I tried playing with escoria and godot 3.1 and it's not possible because of the changes in the gdscript.
Until now you could assign the item.gd which extends interactive.gd extends Node to different node types like Node, Node2D, TextureRect and then dynamically check the type of the node and run a method, but now it is not allowed, unless with hacks, and is also not encouraged. Read on https://github.com/godotengine/godot/issues/22176

**Comment by mjtorn** [comment]
Let's see what the ramifications of this turns out to be, but I think it would be safe - and all-around best - to have only `Node2D` supported. Do you concur?

**Comment by santouits** [comment]
If it were up to me I would make two kind of scripts, one for the inventory items that would extend something from base control, like TextureRect, and one for the world items that would extend KinematicBody2D, so you could change character if you would like and move them, and have the engine's pathfinding.

**Comment by mjtorn** [comment]
Contributions are welcome :)

I think we'd entertain the idea of having a separate inventory item type, as it sometimes feels a bit overloaded to have only `item.gd`. As long as it doesn't break the current inventory logic and @StraToN signs off on it ;) Making it only `TextureRect` might be a bit constraining, though, in case you want animated mouse-overs or something on inventory items.

`KinematicBody2D` is a must for NPCs, but no one's had the time to really care... But it can't be the only kind of item, because some items are static `Area2D`s as part of the background, or maybe a `Sprite`. 

So basically `npc.gd`, `inventory_item.gd` which may or may not be a UI element, and the current `item.gd` which would be adapted to work with `Node2D`.

**Comment by santouits** [comment]
I've been thinking of ways to make it work in 3.1 and I made a prototype, though it has many changes to the current system. 
In every object in godot you can use the set_meta method to put something there. So instead of attaching the item.gd script to a node, we add it as a component of the node. We could also add it as a child, if we would want that, but this is what I chose. All that happens now automatically.

So what I did:

- I took all the code from the interactive.gd and item.gd and put it in the script globals/component.gd
- When the scene is loading, the main.gd iterates the children of the scene and puts that component.gd to them as a meta object and then runs the init method of the component.gd which is the same as running the _ready method of the item.gd.
- Now the developer instead of attaching the item.gd to every item in the scene, they will just create a script and set the global_id and the other variables like before, or they can use the already custom nodes I put in the node creation window by searching "Escoria" there, and they have the exported variables like before.

If you want to test it you can download from here https://github.com/santouits/escoria/releases/tag/0.0001

When you first open with the editor I think the plugin doesn't get enabled for some reason so you should go to project/project settings/plugins and set it to active or restart the editor. Otherwise the custom nodes I added will not show in the window where you create nodes.

**Comment by mjtorn** [comment]
Would you be so kind as to release the code to be able to see a diff? I don't know what kind of opinion to form, but I couldn't open the demo project with 3.0.x.

You didn't mean to say 3.1, did you?

**Comment by santouits** [comment]
Here is the diff https://github.com/santouits/escoria/commit/3f5faa3fc2a6a068d3ce2a3ac69596d11e9f274f

and the code after some changes if you can run it https://github.com/santouits/escoria/releases/tag/0.0002

If you tell me  what errors it shows when running the demo. I use 3.0.6.

**Comment by santouits** [comment]
Oh yeah I didn't bother with room2 yet, I haven't even opened it yet.

If we use only Node2D then every item would have to be that, and I think the flow of developing a game would be harder. Now we just add a component to any node and it is just like before. It's not that a big change I think and I think it is quite clean. The opposite would be to have many scripts for every node type we want to use and the worse with the same script copied all over. Now you just add a script with some variables for the component script to read them.

The exit and trigger logic were inserted inside the component script to demonstrate this logic, so I removed them. You can make an object now triggerable by setting an enum. 

The changes are 3,
- Changed from attaching item.gd to the component.gd
- Changed the demo scene, because it wouldn't run now after the changes
- Added the plugin for adding nodes that don't need to add a script for having exported variables

I think the first two can't be merged alone, or in steps.

**Comment by mjtorn** [comment]
I'll have to look at the implementation when I have more time. Regardless of that, I think the way to structure this is first a branch with one commit:

  1) Add the plugin with current scripts as they are in `master`

I'm pretty sure that would get merged, because we need to make Escoria a proper asset. This has been in the talks since #5 and anything helps ;) I tried making something bigger and totally did not take this baby-step first, so I gave up. This was half a year ago....

Then once that basic groundwork is out of the way, make a branch with commits for the following

  1) Refactor the scripts and plugin
  2) Refactor the demo scenes
  3) Update documentation and write new documentation

Then it's about 512 times easier to review than that diff.

Out of curiosity, are you making a game with this at the moment?

Because Escoria doesn't really come with proper tests and I only know of our [Iron Sky: A Lunar Adventure](https://ironsky.game/) as a recent release. It's hard enough to take that code base every once in a while, and update it onto `master`, and make sure everything works, and possibly release it with hopefully enough testing.

A sweeping change like this will need to get the **REDACTED** tested out of it.

**Comment by santouits** [comment]
You are right of course. No I'm not making a game right now, just playing. I think I'll start by making a propose merging the interactive and trigger scripts inside item.gd, and talk about it.

**Comment by mjtorn** [comment]
By all means, but it's going to be a very tough sell. The reason why `trigger.gd` and `exit.gd` exist is that they are not wanted in `item.gd`. These scrips are very overloaded and there's a ton of stuff even in `interactive.gd` that wouldn't really be necessary for an exit.

But I'd love to see it and talk about it, maybe you've come up with something superior :)

**Comment by santouits** [comment]
So I thought about it a little more and here are some options. The problem with godot 3.1 is that item.gd extends Node and we can add it to any node for now but it will refuse to check and run functions of any node that isn't of type Node.

- We keep attaching the item.gd to every object we need, but the script creates another script that does the checking and the logic. This I guess will not break any projects. But this is so limited. And we don't know if in the future Godot will not allow us to attach scripts of Node to other types like it does now.

- We make a new custom node(I really like this idea) that extends Node and we name it EscoriaObject or something and add it as a child of every node we want to script. It would even have an escoria icon. So the workflow would be make my item with the godot tools and add this escoria object as a child, in contrast to make my item with godo tools and add the item script to the same node. We could even have some more template scripts so the inspector doesn't get spammed with all the exported variables if you want to make let's say a simple trigger. This will break projects, but the work to correct them would be adding the escoria node as a child of an item, copying the exported variables from the parent to that node, removing the item.gd from the parent.

**Comment by mjtorn** [comment]
> We could even have some more template scripts so the inspector doesn't get spammed with all the exported variables if you want to make let's say a simple trigger.

I might be thinking more along the lines of Unity, where you attach many scripts, that not exporting all variables could be made by having `EscoriaObject` or somesuch and having child nodes with "traits" (or whatever you'd call them).

`EscoriaObject` would have the child `EscoriaSprite` which does only event delegation and `TraitItem` (or  `ItemTrait, but having `Trait*` would sort them better) which would give it a tooltip and the ability to pick it up etc. If you want it to talk, but not be a full NPC (ie. it would be unable to move, like Murray the Demonic Skull), you'd add a `TraitTalk` node to it which exposes dialog color and all that.

The subtrees for every `EscoriaObject` would get bigger, but they'd be better separated by concern and have less duplication.

Sadly this is a major overhaul with a lot of question marks about how to manage having an unknown amount of unknown traits, and how to deal with conflicts if something is both a `trigger` and an `exit`, etc etc.

**Comment by santouits** [comment]
Yeah I like this idea with the escoriaobject having children, it is a much cleaner and userfriendly way than different tempates. 
IMO the current work flow of escoria is tedious and can't evolve into something more userfriendly like other adventure frameworks. It also produces bad code when you want to add a new feature. It needs some architecture change, whether it be something like this or something else.
Thankfully the code base is quite small.
The one thing that is a drawback with this system is that Godot 3.1 is slow, and gets a lot slower by adding more children to the tree, especially if they extend Control. This is a godot problem though.

**Comment by mjtorn** [comment]
@santouits do you have benchmarks or something for the slowness? I didn't really see a lot of problems when we were testing 3.1, but that was only PC and it looks like a long time coming for us on mobile platforms.

Fortunately I think we can phase out Control from the actual game UIs, but it'll take time to get started and I might be very busy for a very long time with non-Godot stuff in the near future :(

As for the tediousness, I dunno. The learning curve totally sucks because any tutorials and demos are out of date, but it's not all impossible to get things done here. Someone could argue that adding a bunch of children for traits is tedious as well.

IMO the shitty parts are things like RichTextLabel not doing vertical alignment properly and not having a way to get the width of text inside any type of label, which makes dialog really hard to reason about. Godot problems as well.

**Comment by santouits** [comment]
I don't have any benchmarks, it is just my experience while I'm making my game(not with escoria). I use an old laptop and I have made an inventory using 40 controls, while each one is having control based children and when I made it visible the frame rate fell drastically. It was happening when making the inventory visible and not while adding, but I didn't try to find the reason, just removed the controls, because I don't use the mouse as input anyway in my game.
But in a point and click framework control-based nodes for mouse detection are a must, and they can get too many quite easily if the developer doesn't know it.
Some references:
https://github.com/godotengine/godot/issues/27230
https://godotengine.org/qa/14473/stuttering-and-freezing-when-moving-mouse-on-large-control

**Comment by mjtorn** [comment]
Seems like a "Godot moment", that something that should work just turns to torture for no apparent reason. Maybe it would help to give Patreon money and vote for better performance so maybe it would happen some year. Or just buy into UE4 or even Unity.

But for adventure games, I'm not sure if that's a real issue; there probably wouldn't be as many nodes as in an RTS. Even if you had something like *The Secret of Monkey Island* where you revisit the beginning of the game but characters and things are different, chances are you'd reuse the assets but make separate rooms, as that makes handling the ESC scripts also easier. Maybe, I dunno, we haven't done that and haven't had problems there.

>  But in a point and click framework control-based nodes for mouse detection are a must, and they can get too many quite easily if the developer doesn't know it.

Why is that? Escoria has dropped most of the `Control`-based nodes and things work just fine. Actually the only mouse-related problems come from how the `TextureRect`s capture all input, and there doesn't seem to be a way to tell it "capture clicks, but pass the motion to whatever's underneath". I hope I'm wrong and just haven't looked hard enough.

The way I see it, using `Control`-based nodes is nothing but pain otherwise. What if you want an `AnimatedSprite` for your action menu? You'd currently anchor it in a `Container` and something inheriting from `BaseButton` just to make the `action_menu.gd` script play along. IMO there's little of value in `Container`s here, because you should lay out the action menu manually anyway. And you could listen to click events on the sprites. The only "problem" is that `vm.hover_stack` has to be managed properly, but I'm not sure that's even a real problem, only something I haven't hide time to look at. So all of this feels like paying "API service" to the existing scripts instead of figuring out better ways of doing things.

It's so annoying that when the tooltip follows the mouse, you get these issues where you press an action menu button and see a stale tooltip because of mouse catching.

The even harder thing to cope with `Control`-based crap is hovering an item/npc/whatever, hitting escape to open the in-game menu, moving the mouse to either no item or another item, hitting escape again, and seeing the stale tooltip. That one is a lot harder to fix, because for menus it *actually* makes sense to use `Control`.

What a shitshow.

**Comment by santouits** [comment]
Yeah I guess controls aren't necessary for mouse detection, I'm just being lazy using them. I'm gonna read the code again to see what you talking about the others.

**Comment by girng** [comment]
> Seems like a "Godot moment", that something that should work just turns to torture for no apparent reason.  

Godot 3: hold my tidepod!!

**Comment by StraToN** [comment]
Godot is not perfect, they make their own choices and Escoria has to deal with it.
Anyway, since Escoria is being rewritten, items detection should now work fine.
